[key reuse] don't consume key in fold_in

jakevdp opened the pull request:
Why? We've found in practice that downstream projects use fold_in multiple times with the same key. This is safe so long as the folded-in value is different every time; in this sense fold_in() is similar to seed(), and for now we must trust the user to not repeat seeds.

mattjj reviewed (approved):
Maybe add a test modeling the "fold in multiple times with different seeds/hashes" use case?

jakevdp committed:
[key reuse] don't consume key in fold_in

Why? We've found in practice that downstream projects use fold_in multiple
times with the same key. This is safe so long as the folded-in value is
different every time; in this sense fold_in() is similar to seed(), and
for now we must trust the user to not repeat seeds.

jakevdp commented:
> Maybe add a test modeling the "fold in multiple times with different seeds/hashes" use case?

Great idea - Done!
